chemprop/multitask_utils.py

```diff
@@ -0,0 +1,84 @@
+from typing import List
+import numpy as np
+
+from chemprop.data import MoleculeDataset
+
+
+def reshape_values(
+    values: List[List[List[float]]],
+    test_data: MoleculeDataset,
+    natom_targets: int,
+    nbond_targets: int,
+    num_tasks: int,
+) -> List[List[List[float]]]:
+    """
+    Reshape the input from shape (num_tasks, number of atomic/bond properties for each task, 1)
+    to shape (data_size, num_tasks, number of atomic/bond properties for this data in each task).
+
+    :param values: List of atomic/bond properties with shape
+                   (num_tasks, number of atomic/bond properties for each task, 1).
+    :param test_data: A :class:`~chemprop.data.MoleculeDataset` containing valid datapoints.
+    :param natom_targets: The number of atomic targets.
+    :param nbond_targets: The number of bond targets.
+    :param num_tasks: Number of tasks.
+    :return: List of atomic/bond properties with shape
+             (data_size, num_tasks, number of atomic/bond properties for this data in each task).
+    """
+    n_atoms, n_bonds = test_data.number_of_atoms, test_data.number_of_bonds
+    reshaped_values = []
+    for i in range(natom_targets):
+        reshaped_values.append(
+            np.split(values[i].flatten(), np.cumsum(np.array(n_atoms)))[:-1]
+        )
+    for i in range(nbond_targets):
+        reshaped_values.append(
+            np.split(
+                values[i + natom_targets].flatten(), np.cumsum(np.array(n_bonds))
+            )[:-1]
+        )
+    reshaped_values = [
+        [reshaped_values[j][i] for j in range(num_tasks)] for i in range(len(test_data))
+    ]
+    return reshaped_values
+
+
+def reshape_preds(
+    individual_preds: List[List[List[List[float]]]],
+    test_data: MoleculeDataset,
+    natom_targets: int,
+    nbond_targets: int,
+    num_tasks: int,
+    num_models: int,
+) -> List[List[List[List[float]]]]:
+    """
+    Reshape the input from shape (num_tasks, number of atomic/bond properties for each task, 1, num_models)
+    to shape (data_size, num_tasks, num_models, number of atomic/bond properties for this data in each task).
+
+    :param individual_preds: List of atomic/bond properties with shape
+                             (num_tasks, number of atomic/bond properties for each task, 1, num_models).
+    :param test_data: A :class:`~chemprop.data.MoleculeDataset` containing valid datapoints.
+    :param natom_targets: The number of atomic targets.
+    :param nbond_targets: The number of bond targets.
+    :param num_tasks: Number of tasks.
+    :param num_models: Number of models.
+    :return: List of atomic/bond properties with shape
+             (data_size, num_tasks, num_models, number of atomic/bond properties for this data in each task).
+    """
+    n_atoms, n_bonds = test_data.number_of_atoms, test_data.number_of_bonds
+    individual_values = [[[] for j in range(num_tasks)] for i in range(len(test_data))]
+    for i in range(natom_targets):
```

Comment:
Nested for loops are known to be inefficient. Since `natom_targets`, `nbond_targets`, and `num_models` could potentially be large, the loops here might take a while.

Maybe consider refactoring using numpy built-in functionalities to achieve the desired reshape. If it takes too much time, we can keep this for now. Thanks.

Reply:
That is the thing I should do. I will try to refactor it.

Reply:
@oscarwumit I have refactored those functions. The code should be more efficient now.

Reply:
thanks